share/tools/web_config/sample_prompts/sorin.fish

```diff
@@ -1,61 +1,219 @@
-            switch $i
-                case "."
-                    printf (set_color green)✚' '
-                case " D"
-                    printf (set_color red)✖' '
-                case "*M*"
-                    printf (set_color blue)✱' '
-                case "*R*"
-                    printf (set_color brmagenta)➜' '
-                case "*U*"
-                    printf (set_color bryellow)═' '
-                case "??"
-                    printf (set_color brwhite)◼' '
+    # Count added, deleted, modified, renamed, unmerged, and untracked.
+    # T (type change) is undocumented, see http://git.io/FnpMGw.
+    # For a table of scenarii, see http://i.imgur.com/2YLu1.png.
+    set -l status_added 0
+    set -l status_deleted 0
+    set -l status_modified 0
+    set -l status_renamed 0
+    set -l status_unmerged 0
+    set -l status_untracked 0
+    for line in (command git status --porcelain | cut -b-2)
+        # Check unambiguous cases first which allows us
+        # to skip running all the other regexps.
+        if test "$line" = '??'
+            set status_untracked 1
+            continue
+        end
+        if string match -r '^(?:AA|DD|U.|.U)$' "$line" >/dev/null
+            set status_unmerged 1
+            continue
+        end
+        if string match -r '^(?:[ACDMT][ MT]|[ACMT]D)$' "$line" >/dev/null
+            set status_added 1
+        end
+        if string match -r '^[ ACMRT]D$' "$line" >/dev/null
+            set status_deleted 1
+        end
+        if string match -r '^.[MT]$' "$line" >/dev/null
+            set status_modified 1
+        end
+        if string match -e 'R' "$line" >/dev/null
+            set status_renamed 1
+        end
+    end
+
+    set_color -o
+
+    if test -n "$branch"
+        if test $branch_detached -ne 0
+            set_color brmagenta
+        else
+            set_color green
+        end
+        echo -n " $branch"
+    end
+    if test -n "$commit"
+        echo -n ' '(set_color yellow)"$commit"
+    end
+    if test -n "$action"
+        set_color normal
+        echo -n (set_color white)':'(set_color -o brred)"$action"
+    end
+    if test $status_ahead -ne 0
+        echo -n ' '(set_color brmagenta)'⬆'
+    end
+    if test $status_behind -ne 0
+        echo -n ' '(set_color brmagenta)'⬇'
+    end
+    if test $status_stashed -ne 0
+        echo -n ' '(set_color cyan)'✭'
+    end
+    if test $status_added -ne 0
+        echo -n ' '(set_color green)'✚'
+    end
+    if test $status_deleted -ne 0
+        echo -n ' '(set_color red)'✖'
+    end
+    if test $status_modified -ne 0
+        echo -n ' '(set_color blue)'✱'
+    end
+    if test $status_renamed -ne 0
+        echo -n ' '(set_color magenta)'➜'
+    end
+    if test $status_unmerged -ne 0
+        echo -n ' '(set_color yellow)'═'
+    end
+    if test $status_untracked -ne 0
+        echo -n ' '(set_color white)'◼'
+    end
+
+    set_color normal
+end
+
+function _git_action --argument-names git_dir
```

Comment:
This could probably be extracted as a helper function.

Reply:
I extracted it into `__fish_git_action` now. Do you prefer this?
Or would you maybe like to rename the function?

Reply:
Hmm... I was just thinking about this and...
I believe it might be better to keep it as a private function, because as soon as it isn't anymore, you need to keep backwards-compatibility in mind, right?
And I'm not sure if we should push for that already, since this function is kinda... specialized, unlike `fish_print_hg_root`.

What's your opinion on this?

Reply:
>this function is kinda... specialized, unlike fish_print_hg_root.

Is it? I mean it's mostly translating git's "API" into strings. You'd have to keep those constant for any users, but that's not that huge a burden.

Absolute worst case? You translate them again in the prompt. E.g. we could replace https://github.com/fish-shell/fish-shell/blob/master/share/functions/__fish_git_prompt.fish#L562-L592 with something like

```fish
switch (fish_git_action)
    case rebase
        echo "|REBASE"
    case merge
        echo "|MERGING"
    # and so on
end
```

And you'd still isolate a bunch of knowledge about git in one function.

And that prompt is a nice example, because `git blame -w` tells us that, ignoring whitespace-only changes, those strings haven't changed since 2013.

Reply:
> I named it `fish_print_git_action` to be a bit analogue to your hq command.
> 

I just tried it from today's build and it looks like `fish_right_prompt` is still calling `__fish_git_action`. Once I changed it to `fish_print_git_action`, it worked fine.

Reply:
Fixed in master, sorry!

Reply:
Thanks for taking care of this @faho. 🙂